src/completion_provider.rs

```diff
@@ -0,0 +1,335 @@
+    let ret_ptr = match ret {
+        Some(ref ret) => ret.as_ptr(),
+        None => std::ptr::null_mut(),
+    };
+
+    let old_ptr = gobject_sys::g_object_get_qdata(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_ICON_QUARK.to_glib(),
+    );
+    if !old_ptr.is_null() {
+        assert_eq!(old_ptr as *mut _, ret_ptr, "Did not return same icon again");
+    }
+
+    gobject_sys::g_object_set_qdata_full(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_ICON_QUARK.to_glib(),
+        ret_ptr as *mut libc::c_void,
+        Some(unref),
+    );
+
+    ret.to_glib_none().0
+}
+
+static COMPLETION_PROVIDER_ICON_NAME_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk-sourceview-subclass-icon-name"));
+
+unsafe extern "C" fn completion_provider_get_icon_name<
+    T: ObjectSubclass + CompletionProviderImpl,
+>(
+    completion_provider: *mut gtk_source_sys::GtkSourceCompletionProvider,
+) -> *const libc::c_char {
+    let instance = &*(completion_provider as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let ret = imp.get_icon_name(&from_glib_borrow(completion_provider));
+    let ret_ptr = match ret {
+        Some(ref ret) => ret.as_ptr(),
+        None => std::ptr::null(),
+    };
+
+    let old_ptr = gobject_sys::g_object_get_qdata(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_ICON_NAME_QUARK.to_glib(),
+    );
+    if !old_ptr.is_null() {
+        assert_eq!(
+            old_ptr as *const _, ret_ptr,
+            "Did not return same icon name again"
+        );
+    }
+
+    gobject_sys::g_object_set_qdata_full(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_ICON_NAME_QUARK.to_glib(),
+        ret_ptr as *mut libc::c_void,
+        Some(unref),
+    );
+
+    ret.to_glib_none().0
+}
+
+static COMPLETION_PROVIDER_GICON_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk-sourceview-subclass-gicon"));
+
+unsafe extern "C" fn completion_provider_get_gicon<T: ObjectSubclass + CompletionProviderImpl>(
+    completion_provider: *mut gtk_source_sys::GtkSourceCompletionProvider,
+) -> *mut gio_sys::GIcon {
+    let instance = &*(completion_provider as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let ret = imp.get_gicon(&from_glib_borrow(completion_provider));
+    let ret_ptr = match ret {
+        Some(ref ret) => ret.as_ptr(),
+        None => std::ptr::null_mut(),
+    };
+
+    let old_ptr = gobject_sys::g_object_get_qdata(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_GICON_QUARK.to_glib(),
+    );
+    if !old_ptr.is_null() {
+        assert_eq!(
+            old_ptr as *mut _, ret_ptr,
+            "Did not return same gicon again"
+        );
+    }
+
+    gobject_sys::g_object_set_qdata_full(
+        completion_provider as *mut _,
+        COMPLETION_PROVIDER_GICON_QUARK.to_glib(),
+        ret_ptr as *mut libc::c_void,
```

Comment:
Also here and elsewhere
```suggestion
        ret.to_glib_full() as *mut libc::c_void,
```

And the whole thing only if `ptr` is `Some`

Reply:
See also https://github.com/gtk-rs/gio/pull/299